src/dbnode/storage/shard.go

```diff
@@ -1409,12 +1409,111 @@ func (s *dbShard) insertSeriesSync(
 	return newEntry, nil
 }
 
+func (s *dbShard) insertNewShardEntriesWithLock(entries []*Entry) {
+	if len(entries) == 0 {
+		return
+	}
+
+	// Fast Path: Check if the entire slice can be appended to the end of the list
+	if s.canAppendToEnd(entries[0]) {
+		s.appendEntriesToEnd(entries)
+		return
+	}
+
+	// If not, proceed with the standard insertion logic
+	elem := s.list.Back()
+	i := len(entries) - 1
```

Comment:
hm this code is a little hard to follow. instead of working with indexes like this, if you work with the slice itself - it'll probably be a lot simpler to reason about.

ex. change the for loop condition to something like `for elem != nil && len(entries) > 0 {` and all your helper functions can operate on subslices

Reply:
Refactored the code accordingly